Add support for optimized CMSIS-NN kernels with TFLM

mansnils opened the pull request:
Updates TensorFlow Lite Micro module with option for CMSIS-NN optimized
kernels.
Updates corresponding sample hello_world to use CMSIS-NN optimized kernels.

Signed-off-by: Måns Nilsson <mans.nilsson@arm.com>

stephanosio reviewed (changes requested):
Please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines) and format the commit title and message accordingly.

mansnils committed:
modules: add support for optimized CMSIS-NN kernels

Updates TensorFlow Lite Micro module with option for CMSIS-NN optimized
kernels.

Signed-off-by: Måns Nilsson <mans.nilsson@arm.com>

mansnils committed:
samples: add support for optimized CMSIS-NN kernels

Updates tflite-micro sample hello_world to use CMSIS-NN optimized
kernels. This is also documented in the corresponding README file.

Signed-off-by: Måns Nilsson <mans.nilsson@arm.com>

JordanYates commented on modules/tflite-micro/Kconfig:
```diff
@@ -8,3 +9,24 @@ config TENSORFLOW_LITE_MICRO
+
+if CPU_CORTEX_M
+
+config OPTIMIZED_CMSIS_NN_KERNELS
```
I think this option needs to be clearer about what it is.
Perhaps `TENSORFLOW_LITE_MICRO_CMSIS_NN_KERNELS`

mansnils replied:
Sounds good to me

mansnils replied:
Done

stephanosio commented on modules/tflite-micro/CMakeLists.txt:
```diff
@@ -14,6 +15,14 @@ if(CONFIG_TENSORFLOW_LITE_MICRO)
     ${TENSORFLOW_LITE_MICRO_DIR}/third_party_static/ruy
   )
 
+  if (CONFIG_TENSORFLOW_LITE_MICRO_CMSIS_NN_KERNELS)
```
Nit: spurious space

```suggestion
  if(CONFIG_TENSORFLOW_LITE_MICRO_CMSIS_NN_KERNELS)
```

mansnils replied:
Done

stephanosio commented on modules/tflite-micro/CMakeLists.txt:
```diff
@@ -14,6 +15,14 @@ if(CONFIG_TENSORFLOW_LITE_MICRO)
+    set(CMSIS_NN_OPTIMIZED_KERNEL_DIR cmsis_nn)
+    set(tflm_cmsis_nn_glue_path ${ZEPHYR_CMSIS_MODULE_DIR})
+
+    zephyr_include_directories(${tflm_cmsis_nn_glue_path})
```
assuming this is only required for building the "tflite-micro library," this should use `zephyr_library_include_directories`.

```suggestion
    zephyr_library_include_directories(${tflm_cmsis_nn_glue_path})
```

stephanosio commented on modules/tflite-micro/CMakeLists.txt:
```diff
@@ -14,6 +15,14 @@ if(CONFIG_TENSORFLOW_LITE_MICRO)
+    set(tflm_cmsis_nn_glue_path ${ZEPHYR_CMSIS_MODULE_DIR})
+
+    zephyr_include_directories(${tflm_cmsis_nn_glue_path})
+    add_definitions(-DCMSIS_NN)
```
also assuming `-DCMSIS_NN` is only necessary for building the "tflite-micro library," this should use `zephyr_library_compile_definitions`.

```suggestion
    zephyr_library_compile_definitions(CMSIS_NN)
```

stephanosio commented on samples/modules/tflite-micro/hello_world/prj.conf:
```diff
@@ -17,3 +17,4 @@ CONFIG_LIB_CPLUSPLUS=y
 CONFIG_NEWLIB_LIBC=y
 CONFIG_TENSORFLOW_LITE_MICRO=y
 CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_TENSORFLOW_LITE_MICRO_CMSIS_NN_KERNELS=y
```
Since this sample can run on a target other than Cortex-M, this should not be added to the common configs.

Also note that the `sample.yaml` currently restricts this sample to `qemu_x86` and `qemu_x86_64`:
https://github.com/zephyrproject-rtos/zephyr/blob/fe7e6069a49e4ddbf0d31d299a8c41714bee6904/samples/modules/tflite-micro/hello_world/sample.yaml#L17

I would suggest adding a separate "test" for the CMSIS-NN:
```yaml
tests:
  ...
  sample.tensorflow.helloworld.cmsis_nn:
    tags: tensorflow
    platform_allow: mps3_an547
    extra_configs:
    - CONFIG_TENSORFLOW_LITE_MICRO_CMSIS_NN_KERNELS=y
```

mansnils replied:
Done

carlescufi commented:
@mansnils can you please fix compliance?

![image](https://user-images.githubusercontent.com/12450381/176960681-36bd0e0f-595e-4c40-9edb-94cbfb8b8a12.png)

mansnils commented:
> @mansnils can you please fix compliance?
> 
> ![image](https://user-images.githubusercontent.com/12450381/176960681-36bd0e0f-595e-4c40-9edb-94cbfb8b8a12.png)

Yes will do. Sorry for slow response.
